homeassistant/components/sensor/modem_callerid.py

```diff
@@ -0,0 +1,205 @@
+https://home-assistant.io/components/sensor.modem_callerid/
+"""
+import logging
+import threading
+import datetime
+import voluptuous as vol
+from homeassistant.const import (STATE_IDLE,
+                                 STATE_UNAVAILABLE,
+                                 EVENT_HOMEASSISTANT_STOP,
+                                 CONF_NAME,
+                                 CONF_DEVICE,
+                                 CONF_CODE)
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['pyserial==3.1.1']
+
+_LOGGER = logging.getLogger(__name__)
+DEFAULT_NAME = 'Modem CallerID'
+ICON = 'mdi:phone-clasic'
+DEFAULT_DEVICE = '/dev/ttyACM0'
+DEFAULT_INITSTRING = 'AT#CID=1'
+
+STATE_RING = 'ring'
+STATE_CALLERID = 'callerid'
+
+RING_TIMEOUT = 10
+RING_WAIT = None
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+    vol.Optional(CONF_CODE, default=DEFAULT_INITSTRING): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup modem caller id sensor platform."""
+    name = config.get(CONF_NAME)
+    dev = config.get(CONF_DEVICE)
+    init_string = config.get(CONF_CODE) + '\r\n'
+    add_devices([ModemCalleridSensor(hass, name, dev, init_string)])
+
+
+class ModemCalleridSensor(Entity):
+    """Implementation of USB modem callerid."""
+
+    def __init__(self, hass, name, dev, init_string):
+        """Initialize the sensor."""
+        self._state = STATE_IDLE
+        self._attributes = {"cid_time": 0, "cid_number": '', "cid_name": ''}
+        self._name = name
+        self.dev = dev
+        self.init_string = init_string
+        self.ser = None
+
+        threading.Thread(target=self._modem_sm, daemon=True).start()
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self._stop_modem)
+
+    def set_state(self, state):
+        """Set the state."""
+        self._state = state
+
+    def set_attributes(self, attributes):
+        """Set the state attributes."""
+        self._attributes = attributes
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def icon(self):
+        """Return icon."""
+        return ICON
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    def _stop_modem(self, event):
+        """HA is shutting down, close modem port."""
+        if self.ser:
+            self.ser.close()
+            self.ser = None
+        return
+
+    def _modem_sm(self):
```

Comment:
Hi,

Thanks for the quick feedback.  I based my implementation using other sensors for guidance, attempting to keep it as generic as possible to work with most USB modems.  I can see about extracting the state machine to a 3rd party library, have to give it some thought.  Do you have a recommendation of a good example to base my updates on?

regards,
Tim